packages/flutter_test/lib/src/matchers.dart

```diff
@@ -592,7 +592,7 @@ class _HasGoodToStringDeep extends Matcher {
 /// sets of value for which a metric space is defined.
 typedef num DistanceFunction<T>(T a, T b);
 
-const Map<Type, DistanceFunction<dynamic>> _kStandardDistanceFunctions = const <Type, DistanceFunction<dynamic>>{
+const Map<Type, DistanceFunction<Null>> _kStandardDistanceFunctions = const <Type, DistanceFunction<Null>>{
```

Comment:
this doesn't seem right. They don't return null. They return a variety of values.

Reply:
@Hixie Is this one good?

Reply:
`DistanceFunction<Null>`, to me, reads as "a function that takes null".
I'm pretty sure "null" is by definition not any of the allowed values of these functions. :-)

`dynamic` seems correct: it's saying, "I don't know what this takes".

Reply:
> I'm pretty sure "null" is by definition not any of the allowed values of these functions. :-)

If we had NNBD with a true bottom type (i.e. not inhabited even by null) then yes, bottom would be a better choice here.  But at the moment, null *is* an inhabitant of all types, and all functions are (from the type system perspective) functions that take null.

> dynamic seems correct: it's saying, "I don't know what this takes".

It doesn't mean that though.  It means "I know that these functions are prepared to receive any value", which the functions in this map are not.  Each individual function takes a particular set of values, and the intersection of those sets is `null`.

That said, I'm in agreement that using `Null` for this is very non-intuitive, and it doesn't play all that well with inference.  @lrhn and I have been discussing how we can improve this in Dart 2.x, because being able to say "I don't know what this takes" is useful.  The `dynamic` type basically did mean that in Dart 1.0, since it was treated both as bottom (Null) and as top (Object), but the fact that you could use it anywhere broke a lot of basic things.  Removing this dual role for `dynamic` was probably the most fundamental change in strong mode/Dart 2.0.

Reply:
@Hixie Is this one good?

Reply:
I can't really get used to this. If I say I have a `Future<Null>`, I'm not saying I can return anything. If I have a list of `DistanceFunction<Null>` closures, it's weird to me that I could provide something that isn't null. By weird I mean that I think this is highly unintuitive and makes the code harder to understand.